### PR TITLE
Add completely new interfaces and implementation for FileNamerStrategy

### PR DESCRIPTION
**Pull request by roesslerj**
Separation of concerns: Have NamingStrategy and ProjectLayout separated.

### CODE REVIEW

File: src/main/java/de/retest/recheck/RecheckOptions.java [hunk 1]
```diff
@@ -78,7 +85,10 @@ public String getSuiteName() {
 		if ( suiteName != null ) {
 			return suiteName;
 		}
-		return fileNamerStrategy.getTestClassName();
+		if ( fileNamerStrategy != null ) {
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/RecheckOptions.java]: If this is set, the `fileNamerStrategy` is identical to `namingStrategy`

File: src/main/java/de/retest/recheck/RecheckOptions.java [hunk 2]
```diff
@@ -124,14 +153,28 @@ private File getSuitePath() {
...
 		 *            The strategy to use, defaults to {@link MavenConformFileNamerStrategy}
 		 * @return self
 		 */
+		@Deprecated
 		public RecheckOptionsBuilder fileNamerStrategy( final FileNamerStrategy fileNamerStrategy ) {
 			this.fileNamerStrategy = fileNamerStrategy;
+			return namingStrategy( fileNamerStrategy );
+		}
+
+		public RecheckOptionsBuilder namingStrategy( final NamingStrategy namingStrategy ) {
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/RecheckOptions.java:182]: Javadoc

File: src/main/java/de/retest/recheck/persistence/ExplicitMutableNamingStrategy.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
+package de.retest.recheck.persistence;
+
+public class ExplicitMutableNamingStrategy implements NamingStrategy {
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/persistence/ExplicitMutableNamingStrategy.java:10]: Annotation from lombok with javadoc

File: src/main/java/de/retest/recheck/persistence/GradleProjectLayout.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+	/**
+	 * Creates a Gradle-conform file namer strategy using <code>sourceSetName</code> for
+	 * <code>${SOURCE_SET_NAME}</code>.
+	 *
+	 * @param sourceSetName
+	 *            value for <code>${SOURCE_SET_NAME}</code>
+	 */
+	public GradleProjectLayout( final String sourceSetName ) {
+		super( Paths.get( format( DEFAULT_RETEST_PROJECT_PATH_FORMAT, validate( sourceSetName ) ),
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/persistence/GradleProjectLayout.java]: Extract static methods

File: src/main/java/de/retest/recheck/persistence/JUnitTestbasedShortNamingStrategy.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+	/**
+	 * @return The {@link Class#getSimpleName()} of the test class (e.g. `MyTest`).
+	 */
+	@Override
+	public String getSuiteName() {
+		String result = super.getSuiteName();
+		if ( result.contains( "." ) ) {
+			result = result.substring( result.lastIndexOf( "." ) + 1 );
+		}
+		if ( result.contains( "$" ) ) {
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/persistence/JUnitTestbasedShortNamingStrategy.java]: Do we need to handle inner classes? Is this possible ...
Don't you still expect something like: `InnerClass`

```java
public static class Foo {

	public static class Bar {}
}

Foo.Bar f = new Foo.Bar();
System.out.println(f.getClass().getSimpleName());
```

Output: Bar

File: src/main/java/de/retest/recheck/persistence/SeparatePathsProjectLayout.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+		this.reportPath = reportPath;
+	}
+
+	@Override
+	public Path getGoldenMaster( final String suitename, final String testname, final String checkname ) {
+		String fileName = testname + "." + checkname;
+		if ( StringUtils.isEmpty( testname ) ) {
+			fileName = checkname;
+		}
+		return getSuitsFolder( suitename ).resolve( Paths.get( fileName + GOLDEN_MASTER_FILE_EXTENSION ) );
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/persistence/SeparatePathsProjectLayout.java]: Why not just
```suggestion
		return getSuitsFolder( suitename ).resolve( fileName + GOLDEN_MASTER_FILE_EXTENSION );
```

File: src/main/java/de/retest/recheck/persistence/SeparatePathsProjectLayout.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+	}
+
+	@Override
+	public Path getReport( final String suitename ) {
+		return reportPath.resolve( Paths.get( suitename + TEST_REPORT_FILE_EXTENSION ) );
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/persistence/SeparatePathsProjectLayout.java]: Same here

File: src/main/java/de/retest/recheck/persistence/SeparatePathsProjectLayout.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+	}
+
+	@Override
+	public Path getSuitsFolder( final String suitename ) {
+		return goldenMasterPath.resolve( Paths.get( suitename ) );
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/persistence/SeparatePathsProjectLayout.java]: Same here

File: src/main/java/de/retest/recheck/persistence/JUnitTestbasedShortNamingStrategy.java [hunk 9]
```diff
@@ -0,0 +1,23 @@
+package de.retest.recheck.persistence;
+
+/**
+ * Extends the standard {@link JUnitTestbasedNamingStrategy}, but instead of the qualified class name (e.g.
+ * `com.mycompany.MyTest`) returns the {@link Class#getSimpleName()} of the class (e.g. `MyTest`).
+ */
+public class JUnitTestbasedShortNamingStrategy extends JUnitTestbasedNamingStrategy {
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/persistence/JUnitTestbasedShortNamingStrategy.java]:
```suggestion
public class JUnitTestBasedShortNamingStrategy extends JUnitTestbasedNamingStrategy {
```

File: src/main/java/de/retest/recheck/persistence/JUnitTestbasedNamingStrategy.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
+package de.retest.recheck.persistence;
+
+import de.retest.recheck.TestCaseFinder;
+
+/**
+ * The default implementation that derives test suite and test case name from the test class and test method, working
+ * for both JUnit and TestNG.
+ */
+public class JUnitTestbasedNamingStrategy implements NamingStrategy {
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/persistence/JUnitTestbasedNamingStrategy.java]:
```suggestion
public class JUnitTestBasedNamingStrategy implements NamingStrategy {
```

File: src/main/java/de/retest/recheck/persistence/JUnitTestbasedNamingStrategy.java [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+	/**
+	 * Determines the suite name.
+	 *
+	 * @return name of the current test suite
+	 */
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/persistence/JUnitTestbasedNamingStrategy.java]: Inherit javadoc?

File: src/main/java/de/retest/recheck/persistence/ProjectLayout.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+	/**
+	 * Gets the base-folder of the suite where the tests are stored in. Useful to store additional data, like a
+	 * suite-specific ignore file.
+	 *
+	 * @param suitename
+	 *            The name of the suite.
+	 * @return The base-folder for the given suite name.
+	 */
+	Path getSuitsFolder( final String suitename );
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/persistence/ProjectLayout.java]:
```suggestion
	Path getSuiteFolder( final String suiteName );
```

File: src/main/java/de/retest/recheck/persistence/ProjectLayout.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+	 *
+	 * @param suitename
+	 *            The name to use for the suite to construct the path for the file.
+	 * @param testname
+	 *            The name to use for the test to construct the path for the file.
+	 * @param checkname
+	 *            The name to use for the check to construct the path for the file.
+	 * @return The path to use for storing or retrieving the Golden Master with the given params.
+	 */
+	Path getGoldenMaster( final String suitename, final String testname, final String checkname );
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/persistence/ProjectLayout.java]:
```suggestion
	Path getGoldenMaster( final String suiteName, final String testName, final String checkName );
```

File: src/main/java/de/retest/recheck/persistence/ProjectLayout.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+	/**
+	 * Get the result file for the current suite, which this interface makes no assumptions about how it is contrived.
+	 *
+	 * @param suitename
+	 *            The name to use for the suite to construct the path for the file.
+	 * @return The path to use for storing or retrieving the report with the given param.
+	 */
+	Path getReport( final String suitename );
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/persistence/ProjectLayout.java]:
```suggestion
	Path getReport( final String suiteName );
```

File: src/test/java/de/retest/recheck/RecheckImplTest.java [hunk 15]
```diff
@@ -148,50 +148,22 @@ public boolean canCheck( final Object arg0 ) {
...
-				public File getFile( final String extension ) {
-					return resolveRoot( baseNames, extension );
-				}
-
-				@Override
-				public File getResultFile( final String extension ) {
-					return resolveRoot( baseNames, extension );
-				}
-			};
+		public WithinTempDirectoryProjectLayout( final Path root ) throws IOException {
```
Comment (diba1013) [inline src/test/java/de/retest/recheck/RecheckImplTest.java:152]: While this name is more descriptive, wouldn't it make more sense to add this constructor to `SeparatePathsProjectLayout ` ?

File: src/main/java/de/retest/recheck/persistence/SeparatePathsProjectLayout.java [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Uses two separate paths (provided in the constructor) to place the Golden Masters in a (usually git-managed) resource
+ * folder, but the reports in a temporary build-artefacts folder.
+ *
+ * The naming of the Golden Masters follows <code>${TEST_CLASS_NAME}/${TEST_METHOD_NAME}.${STEP_NAME}</code>, whereas
+ * test report files only use <code>${TEST_CLASS_NAME}</code>.
+ */
+public class SeparatePathsProjectLayout implements ProjectLayout {
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/persistence/SeparatePathsProjectLayout.java:17]:
```suggestion
public class WithinBaseFolderProjectLayout implements ProjectLayout {
```

**Review by diba1013 — CHANGES_REQUESTED** [review]

File: src/main/java/de/retest/recheck/persistence/JUnitTestbasedShortNamingStrategy.java (hunk 5 above)
Comment (roesslerj) [reply src/main/java/de/retest/recheck/persistence/JUnitTestbasedShortNamingStrategy.java]: Just for naming the checks, I would ignore inner classes. Should be a corner case anyways...

Comment (roesslerj) [reply src/main/java/de/retest/recheck/persistence/JUnitTestbasedShortNamingStrategy.java]: If someone wants to change that, they can create a PR.

File: src/main/java/de/retest/recheck/persistence/ProjectLayout.java (hunk 12 above)
Comment (roesslerj) [reply src/main/java/de/retest/recheck/persistence/ProjectLayout.java]: Good catch

File: src/test/java/de/retest/recheck/RecheckImplTest.java (hunk 15 above)
Comment (roesslerj) [reply src/test/java/de/retest/recheck/RecheckImplTest.java:152]: You mean a `SamePaths` ctor in the `SeparatePathsProjectLayout`? :-)

File: src/main/java/de/retest/recheck/persistence/SeparatePathsProjectLayout.java (hunk 16 above)
Comment (roesslerj) [reply src/main/java/de/retest/recheck/persistence/SeparatePathsProjectLayout.java:17]: Hmm. I would argue that he main feature of this class is to have the GoldenMaster and the reports in two distinct folders. Hence the name...

**Review by diba1013 — CHANGES_REQUESTED** [review]
There are still some comments open (in my view there were hidden). Also the build fails.

File: src/main/java/de/retest/recheck/persistence/SeparatePathsProjectLayout.java (hunk 6 above)
Comment (roesslerj) [reply src/main/java/de/retest/recheck/persistence/SeparatePathsProjectLayout.java]: Sehr gut! :-)

File: src/main/java/de/retest/recheck/persistence/JUnitTestbasedNamingStrategy.java (hunk 10 above)
Comment (roesslerj) [reply src/main/java/de/retest/recheck/persistence/JUnitTestbasedNamingStrategy.java]: I had this conversation with Daniel once. We agreed to accept whatever comes out of https://capitalizemytitle.com/camel-case/ ... in our case for "JUnit-test-based naming strategy". After seeing that, I shortened it to "JUnit-based naming strategy"...

File: src/main/java/de/retest/recheck/persistence/JUnitTestbasedShortNamingStrategy.java (hunk 9 above)
Comment (roesslerj) [reply src/main/java/de/retest/recheck/persistence/JUnitTestbasedShortNamingStrategy.java]: Same here

File: src/main/java/de/retest/recheck/persistence/JUnitTestbasedNamingStrategy.java (hunk 11 above)
Comment (roesslerj) [reply src/main/java/de/retest/recheck/persistence/JUnitTestbasedNamingStrategy.java]: Or remove.

File: breaking-changes-2.0.md [hunk 17]
```diff
@@ -0,0 +1,6 @@
+# List of breaking changes for recheck 2.0
```
Comment (diba1013) [inline breaking-changes-2.0.md]: Wouldn' t this be best put into the wiki?

**Review by diba1013 — CHANGES_REQUESTED** [review]

File: breaking-changes-2.0.md (hunk 17 above)
Comment (beatngu13) [reply breaking-changes-2.0.md]: Do we need this list at all? I would suggest to use issues and create a label like `recheck 2.0.0`. Our `@Deprecated` markers should help here.

Comment (roesslerj) [reply breaking-changes-2.0.md]: Ok. I'll remove this from the PR and we can solve it separately...

**Review by diba1013 — APPROVED** [review]

**Pull request merged**